src/main/java/io/github/moulberry/notenoughupdates/recipes/Ingredient.java

```diff
@@ -34,11 +37,19 @@ public class Ingredient {
 	public static final String SKYBLOCK_COIN = "SKYBLOCK_COIN";
 	private final double count;
 	private final String internalItemId;
-	private final NEUManager manager;
 	private ItemStack itemStack;
 
-	public Ingredient(NEUManager manager, String ingredientIdentifier) {
-		this.manager = manager;
+	private static final Map<String, Ingredient> itemCache = new HashMap<>();
+	private static final Map<Integer, Ingredient> itemCoinCache = new HashMap<>();
+	private static final TwoKeyCache<String, Double, Ingredient> itemAmountCache = new TwoKeyCache<>();
```

Comment:
This should probably be using a weak reference map. Something along the lines of 
```suggestion
	private static final Map<Pair<String, Double>, WeakReference<Ingredient>> itemAmountCache = new HashMap<>();
```
with maybe a ReferenceQueue even. Then again since the NEU repo doesnt really ever lose ingredient types, this is maybe not needed to be ever cleaned out? An intentional memory leak of sorts.

Reply:
not sure i have implemented the weak reference correctly

Reply:
You have not. You have only implemented it for one field, completely ignored what references do (they evaporate when garbage collected, so your cached values can turn invalid over time without you doing anything) and didnt implement reference queues.

Reply:
Also why are those not all just one map?

Reply:
> You have not. You have only implemented it for one field, completely ignored what references do (they evaporate when garbage collected, so your cached values can turn invalid over time without you doing anything) and didnt implement reference queues.

How to do it with weak references? Never worked with them, and your explanation does not help me

Would implementing a guava cache wrapper be smarter? I'm thinking of something like SkyHanni's TimeLimitedCache

Reply:
> Also why are those not all just one map?

I have not thought about the difference between Integer for itemCoinCache, and Double for itemAmountCache. But sure, we could cast int to double and put all in one map.